collector/set.go

```diff
@@ -0,0 +1,110 @@
+package collector
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type SetConfig struct {
+	Collectors []Interface
+	Logger     micrologger.Logger
+}
+
+type Set struct {
+	collectors []Interface
+	logger     micrologger.Logger
+
+	bootedCounter uint32
+	mutex         sync.Mutex
+}
+
+func NewSet(config SetConfig) (*Set, error) {
+	if len(config.Collectors) == 0 {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Collectors must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	s := &Set{
+		collectors: config.Collectors,
+		logger:     config.Logger,
+
+		bootedCounter: 0,
+		mutex:         sync.Mutex{},
+	}
+
+	return s, nil
+}
+
+func (s *Set) Boot(ctx context.Context) error {
```

Comment:
For my learning, why is this separated from `NewSet`? To have better control over _when_ it's registered?

Reply:
The constructor we usually call something like `New` should actually only initialize the structure and provide some useful implementation you can work with right away. All other procedures, which may be long running, should be done in a separate step. We call this `Boot`. There may be cases, like in testing situations, in which you only want the thing without it doing a lot of magic.

Reply:
How about:

```
func (s *Set) Boot(ctx context.Context) error {
	var err error
	s.bootOnce.Do(func() {
		err = s.boot(ctx)
	})

	return err
}

func (s *Set) boot(ctx context.Context) error {
	// ...
}
```

Reply:
With `sync.Once` we have a scope which we have to work around for the error. The same goes for goroutines. We do not need either of it here usually. Once being in such a scope the action might rather be asynchronous and cannot be returned anyway. I think it is easier and cleaner with the synchronous approach without scope.

Reply:
This approach is less error prone IMO. `sync.Once` is battle tested. Sync stuff is error prone in general. `sync.Once` is synchronous - this is known. Just saying.

Reply:
I actually took the implementation of `sync.Once` and used it here to get what I want. So this is just fine. We have the battle tested code.